File: prototype/rails/src/test/domain/payments/card_decision_test.rb
require "test_helper"

module Domain
  module Payments
    class CardDecisionTest < ActiveSupport::TestCase
      test "an approved decision carries no reason" do
        decision = CardDecision.approved("4242")

        assert_predicate decision, :approved?
        assert_nil decision.decline_reason
        assert_equal "4242", decision.last_four
      end

      test "a declined decision carries the reason" do
        decision = CardDecision.declined("0002", "generic_decline")

        refute_predicate decision, :approved?
        assert_equal "generic_decline", decision.decline_reason
      end
    end
  end
end

File: prototype/node/src/app/core/payments/card-decision.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { approvedCard, declinedCard, isCardApproved } from './card-decision.ts'

test('an approved decision carries no reason', () => {
  const decision = approvedCard('4242')

  assert.equal(isCardApproved(decision), true)
  assert.equal(decision.declineReason, null)
  assert.equal(decision.lastFour, '4242')
})

test('a declined decision carries the reason', () => {
  const decision = declinedCard('0002', 'generic_decline')

  assert.equal(isCardApproved(decision), false)
  assert.equal(decision.declineReason, 'generic_decline')
})
